nixos/tests/postgresql-upgrade.nix

```diff
@@ -0,0 +1,144 @@
+{ system ? builtins.currentSystem
+, config ? { }
+, pkgs ? import ../.. { inherit system config; }
+}:
+
+let
+  lib = pkgs.lib;
+
+  inherit (import ../lib/testing-python.nix { inherit system pkgs; })
+    makeTest;
+
+  pgPackages = lib.filterAttrs (_: p: !p.jitSupport)
+    (import ../../pkgs/servers/sql/postgresql pkgs);
+
+  end = lib.last (lib.mapAttrsToList lib.nameValuePair pgPackages);
+
+  startPackages = lib.attrsets.removeAttrs pgPackages [ end.name ];
+
+  mkInsert = package: pkgs.writeShellApplication {
+    name = "insert";
+    runtimeInputs = [ package pkgs.gnugrep ];
+    text = ''
+      run_sql() {
+        su -c 'psql testdb -At -v ON_ERROR_STOP=1' postgres
+      }
+
+      run_sql <<- 'QUERY'
+        CREATE TABLE books (
+          title   text,
+          author  text,
+          year    int
+        );
+
+        INSERT INTO books (title, author, year) VALUES
+        ('Do Androids Dream Of Electric Sheep?', 'Philip K. Dick', 1968),
+        ('Neuromancer', 'William Gibson', 1984),
+        ('Cryptonomicon', 'Neal Stephenson', 1999),
+        ('Accelerando', 'Charles Stross', 2005);
+      QUERY
+    '';
+  };
+
+  verify = pkgs.writeShellApplication {
+    name = "verify";
+    runtimeInputs = [ end.value pkgs.gnugrep ];
+    text = ''
+      run_sql() {
+        su -c 'psql testdb -At -v ON_ERROR_STOP=1' postgres
+      }
+
+      die() {
+        echo "$1" >&2 && exit 1
+      }
+
+      run_sql <<- 'QUERY' | grep 't' || die 'No author found 1'
+        SELECT exists(SELECT * FROM books WHERE author = 'Philip K. Dick');
+      QUERY
+
+      run_sql <<- 'QUERY' | grep 't' || die 'No title found'
+        SELECT exists(SELECT * FROM books WHERE title = 'Neuromancer');
+      QUERY
+
+      run_sql <<- 'QUERY' | grep 't' || die 'No year found'
+        SELECT exists(SELECT * FROM books WHERE year = 1999);
+      QUERY
+
+      run_sql <<- 'QUERY' | grep 't' || die 'No author found 2'
+        SELECT exists(SELECT * FROM books WHERE author = 'Charles Stross');
+      QUERY
+    '';
+  };
+
+  mkTest = start: makeTest rec {
+    name = "postgresql-upgrade-${start.psqlSchema}-to-${end.value.psqlSchema}";
+
+    meta.maintainers = [ lib.maintainers.jsoo1 ];
+
+    nodes = {
+      start = {
+        services.postgresql = {
+          enable = true;
+
+          package = start;
+
+          dataDir = "/var/lib/postgresql/${start.psqlSchema}";
+
+          ensureDatabases = [ "testdb" ];
+        };
+      };
+      end = {
+        services.postgresql = {
+          enable = true;
+
+          package = end.value;
+
+          dataDir = "/var/lib/postgresql/${end.value.psqlSchema}";
+
+          upgradeFrom.package = start;
+
+          upgradeFrom.settings = {
+            link = true;
+            # Test listOf settings
+            old-options = lib.cli.toGNUCommandLine { } {
+              d = 3;
```

Comment:
Would you mind adding an explanatory comment what this is precisely doing (i.e. the flag itself here)?